core/src/main/scala/org/broadinstitute/dsde/rawls/entities/EntityService.scala

```diff
@@ -339,35 +344,156 @@ class EntityService(protected val ctx: RawlsRequestContext,
+                               fetchSize = dataSource.dataAccess.fetchSize
+      )
+
+    // translate the Slick stream to a Source
+    Source.fromPublisher(dataSource.database.stream(allAttrsStream))
+  }
+
+  /**
+    * Given a Source containing entity attributes, scroll through that source and combine
+    * attributes into entities. Emit a Source of entities.
+    * <p>
+    * IMPORTANT: this !!requires!! that the incoming Source of attributes is ordered by
+    * entity ID. If the incoming source is not properly ordered, this method will emit
+    * incomplete/duplicate entities.
+    * <p>
+    * Only used internally by EntityService.listEntities, but public to support unit testing
+    * @param dbSource the Source of attributes, typically from a database stream
+    * @return a Source of entities constructed from the attributes
+    */
+  def gatherEntities(dbSource: Source[EntityAndAttributesResult, NotUsed]): Source[Entity, NotUsed] = {
+    // interim classes used while iterating through the stream, allows us to accumulate attributes
+    // until ready to emit an entity
+    trait AttributeStreamElement
+    case class AttrAccum(accum: Seq[EntityAndAttributesResult], entity: Option[Entity]) extends AttributeStreamElement
+    case object EmptyElement extends AttributeStreamElement
+
+    /*
+     * Given the previous and current stream elements, which are produced by the EntityCollector,
+     * calculate the AttrAccum to be output
+     */
+    def gatherOrOutput(previous: AttributeStreamElement, current: AttributeStreamElement): AttrAccum = {
+      // utility function called when an entity is finished or when the stream is finished
+      def entityFinished(prevAttrs: Seq[EntityAndAttributesResult], nextAttrs: Seq[EntityAndAttributesResult]) = {
+        val unmarshalled = dataSource.dataAccess.entityQuery.unmarshalEntities(prevAttrs)
+        // safety check - did the attributes we gathered all marshal into a single entity?
+        if (unmarshalled.size != 1)
+          throw new DataEntityException(s"gatherOrOutput expected only one entity, found ${unmarshalled.size}")
+        AttrAccum(nextAttrs, Some(unmarshalled.head))
+      }
+
+      // inspect the variations of previous and current
+      (previous, current) match {
+        // if both previous and current are empty, it means no entities found
+        case (EmptyElement, EmptyElement) =>
+          AttrAccum(Seq(), None)
+
+        // if previous is empty but current is not, it's the first element
+        case (EmptyElement, curr: AttrAccum) =>
+          curr
+
+        // midstream, we notice that the current entity is the same as the previous entity.
+        // keep gathering attributes for this entity, and don't emit an entity yet.
+        case (prev: AttrAccum, curr: AttrAccum) if prev.accum.head.entityRecord.id == curr.accum.head.entityRecord.id =>
+          val newAccum = prev.accum ++ curr.accum
```

Comment:
Not sure if `:::` would be preferred in terms of concatenation?

(Anyways -- an interesting read: https://stackoverflow.com/questions/6559996/scala-list-concatenation-vs)